airflow/jobs/scheduler_job.py

```diff
@@ -480,13 +490,40 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session) -
                                 " this task has been reached.",
                                 task_instance,
                             )
-                            starved_tasks.add((task_instance.dag_id, task_instance.task_id))
```

Comment:
I think that instead of modifying the `starved_tasks`, we should add a new variable, that keeps track of the `(dag_id, run_id, task_id)` triples. 

When using the `task_concurrency_limit`, then the current behaviour is that all runs are filtered out, but after your change the filtering is done one dagrun at time and it could take more iterations to get to the TIs, that can be queued.